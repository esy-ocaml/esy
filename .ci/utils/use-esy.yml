# steps to install esy globally

steps:
- script: "yarn global add @esy-nightly/esy@0.6.8-bbdfb8"
  displayName: "Install esy (via yarn)"
  condition: ne(variables['AGENT.OS'], 'Windows_NT')
  
- script: |
    yarn config set prefix C:\
    yarn global add @esy-nightly/esy@0.6.8-bbdfb8
  displayName: "Install esy (via yarn)"
  condition: eq(variables['AGENT.OS'], 'Windows_NT')

- bash: |
    YARN_BIN_DIR=$(yarn global bin)
    echo $YARN_BIN_DIR
    echo "##vso[task.setvariable variable=YARN_BIN_DIR]$YARN_BIN_DIR"
  displayName: "Setting $YARN_BIN_DIR"

- bash: |
    echo "##vso[task.setvariable variable=PATH]$PATH:$YARN_BIN_DIR" 
  displayName: "Ensure yarn install global tools are available on $PATH"
  condition: ne(variables['AGENT.OS'], 'Windows_NT')

- powershell: Write-Host "##vso[task.setvariable variable=PATH;]${env:PATH};${env:YARN_BIN_DIR}"
  displayName: "Ensure yarn install global tools are available on $PATH"
  condition: eq(variables['AGENT.OS'], 'Windows_NT')
