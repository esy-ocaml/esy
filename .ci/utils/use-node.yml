# steps to use node on agent

steps:
  - task: NodeTool@0
    displayName: "Use Node 15.x"
    inputs:
      versionSpec: 15.x
