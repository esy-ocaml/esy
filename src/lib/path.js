/**
 * @flow
 */

import invariant from 'invariant';
const path = require('path');

/**
 * Absolute path.
 */
export opaque type AbsolutePath: ConcretePath = string;

/**
 * Path with no `.` or `..` or any bash variables in it.
 */
export opaque type ConcretePath: Path = string;

/**
 * Path which can contain bash/make variables in it, cannot be used in
 * filesystem operations.
 */
export opaque type AbstractPath: Path = string;

export opaque type Path: string = AbsolutePath | ConcretePath | AbstractPath;

export function absolute(p: string): AbsolutePath {
  p = sanitizeAbsolutePath(p);
  return p;
}

export function concrete(p: string): ConcretePath {
  p = sanitizeConcretePath(p);
  return p;
}

export function abstract(p: string): AbstractPath {
  return p;
}

export function join<B: Path | string>(base: B, ...p: string[]): B {
  return (path.join(base, ...p): any);
}

export function basename<P: Path | string>(p: P): string {
  return path.basename(p);
}

export function dirname<P: Path | string>(p: P): P {
  return (path.dirname(p): any);
}

export function length(p: ConcretePath): number {
  return p.length;
}

export function toString(p: Path | string): string {
  return p;
}

export const sep: typeof path.sep = path.sep;

export const relative: typeof path.relative = path.relative;

function sanitizeAbsolutePath(p) {
  p = sanitizeConcretePath(p);
  invariant(path.isAbsolute(p), 'Should be absolute path but got: %s', p);
  return p;
}

function sanitizeConcretePath(p) {
  // NOTE: regexpes are stateful, this is why we are creating them here.
  // Consider resetting their state after matching instead.
  const REMOVE_TRAILING_SLASH_RE = /\/+$/g;
  const ENVIRONMENT_VAR_RE = /\$[a-zA-Z_]/g;
  const DOT_PATH_SEGMENT_RE = /\/\.\//g;
  const DOT_DOT_PATH_SEGMENT_RE = /\/\.\.\//g;
  const SANITIZE_SLASH_RE = /\/+/g;

  invariant(
    DOT_PATH_SEGMENT_RE.exec(p) == null,
    'Should not contain "." segments: %s',
    p,
  );
  invariant(
    DOT_DOT_PATH_SEGMENT_RE.exec(p) == null,
    'Should not contain ".." segments: %s',
    p,
  );
  invariant(
    ENVIRONMENT_VAR_RE.exec(p) == null,
    'Path should not contain variable references: %s',
    p,
  );
  p = path.normalize(p);
  p = p.replace(REMOVE_TRAILING_SLASH_RE, '');
  p = p === '' ? '/' : p;
  return p;
}
