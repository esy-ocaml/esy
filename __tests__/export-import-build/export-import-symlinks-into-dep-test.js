/**
 * @flow
 */

import * as path from 'path';
import {defineTestCaseWithShell} from '../utils';

defineTestCaseWithShell(
  path.join(__dirname, 'fixtures', 'symlinks-into-dep'),
  `
    run esy build

    # package "subdep" should be visible in all envs
    assertStdout "esy subdep" "subdep"
    assertStdout "esy b subdep" "subdep"
    assertStdout "esy x subdep" "subdep"

    # same for package "dep" but it should reuse the impl of "subdep"
    assertStdout "esy dep" "subdep"
    assertStdout "esy b dep" "subdep"
    assertStdout "esy x dep" "subdep"

    # and root package links into "dep" which links into "subdep"
    assertStdout "esy x symlinks-into-dep" "subdep"

    # check that link is here
    storeTarget=$(readlink ../esy/3/i/dep-1.0.0/bin/dep)
    if [[ "$storeTarget" != $ESY_TEST__PREFIX* ]]; then
      failwith "invalid symlink target: $storeTarget"
    fi

    # export build from store
    run esy export-build ../esy/3/i/dep-1.0.0
    cd _export
    run tar xzf dep-1.0.0.tar.gz
    cd ../

    # check symlink target for exported build
    exportedTarget=$(readlink _export/dep-1.0.0/bin/dep)
    if [[ "$exportedTarget" != ________* ]]; then
      failwith "invalid symlink target: $exportedTarget"
    fi

    # drop & import
    run rm -rf ../esy/3/i/dep-1.0.0
    run esy import-build ./_export/dep-1.0.0.tar.gz

    # check symlink target for imported build
    importedTarget=$(readlink ../esy/3/i/dep-1.0.0/bin/dep)
    if [[ "$importedTarget" != $ESY_TEST__PREFIX* ]]; then
      failwith "invalid symlink target: $exportedTarget"
    fi

  `,
  {snapshotExecutionTrace: true},
);
