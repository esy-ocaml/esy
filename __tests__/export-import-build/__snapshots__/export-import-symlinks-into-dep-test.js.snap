// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build symlinks-into-dep 1`] = `
"RUNNING: esy config ls
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
RUNNING: esy build
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin needRebuildTarget <sandbox>/node_modules/.cache/_esy/build/bin/build
esy:bin trying to acquire lock
esy:bin:esyBuild execute
esy:bin trying to acquire lock
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:build-dependencies checking if dependencies are built
RUNNING: esy subdep
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin needRebuildTarget <sandbox>/node_modules/.cache/_esy/build/bin/build
RUNNING: esy b subdep
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin needRebuildTarget <sandbox>/node_modules/.cache/_esy/build/bin/build
esy:bin trying to acquire lock
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:build-dependencies checking if dependencies are built
RUNNING: esy x subdep
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin needRebuildTarget <sandbox>/node_modules/.cache/_esy/build/bin/build
esy:bin trying to acquire lock
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:build-dependencies checking if dependencies are built
RUNNING: esy dep
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin needRebuildTarget <sandbox>/node_modules/.cache/_esy/build/bin/build
RUNNING: esy b dep
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin needRebuildTarget <sandbox>/node_modules/.cache/_esy/build/bin/build
esy:bin trying to acquire lock
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:build-dependencies checking if dependencies are built
RUNNING: esy x dep
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin needRebuildTarget <sandbox>/node_modules/.cache/_esy/build/bin/build
esy:bin trying to acquire lock
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:build-dependencies checking if dependencies are built
RUNNING: esy x symlinks-into-dep
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin needRebuildTarget <sandbox>/node_modules/.cache/_esy/build/bin/build
esy:bin trying to acquire lock
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:build-dependencies checking if dependencies are built
RUNNING: esy export-build ../esy/3/i/dep-1.0.0
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
RUNNING: tar xzf dep-1.0.0.tar.gz
RUNNING: rm -rf ../esy/3/i/dep-1.0.0
RUNNING: esy import-build ./_export/dep-1.0.0.tar.gz
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy"
`;
